nav2_lifecycle/include/nav2_lifecycle/lifecycle_helper_interface.hpp

```diff
@@ -0,0 +1,41 @@
+// Copyright (c) 2019 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef NAV2_LIFECYCLE__LIFECYCLE_HELPER_INTERFACE_HPP_
+#define NAV2_LIFECYCLE__LIFECYCLE_HELPER_INTERFACE_HPP_
+
+#include "rclcpp_lifecycle/lifecycle_node.hpp"
+
+namespace nav2_lifecycle
+{
+
+using CallbackReturn = rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn;
+
+// A lifecycle interface minus onError and onShutdown, which are only relevant for nodes. This
+// has been proposed for rclcpp: https://github.com/ros2/rclcpp/issues/654 and hopefully can be
+// added as rclcpp_lifecycle::LifecycleHelperInterface. If not, we can remove this interface
+// and just use the whole rclcpp_lifecycle::LifecycleNodeInterface.
+
```

Comment:
Would it be better to derive this class from LifecycleNodeInterface and use the delete function syntax (foo() = delete;) to delete the functions we don't want called? Then if the upstream interface is updated, we don't have to change this to match, the changes will be inherited. It could make maintaining this easier, if we ever expect that to change.

Reply:
We can do this, but I was looking for some feedback first about whether it would be useful to have a LifecycleHelperInterface to make the concept explicit. It didn't seem very elegant to have helper classes derive from the LifecycleNodeInterface and always stub out the two functions that don't apply. Not a big deal though.

Reply:
I suggest we leave this LifecycleNodeInterface class here for now 'til we get some feedback from the community. I'll then adapt to the agreed-upon direction.

Reply:
I don't feel too strongly about it, but it's possible `on_error` and `on_shutdown` could still be useful for an arbitrary "helper". For example, both callbacks might do some reporting and/or resource cleanup. 
Since the [default implementation](https://github.com/ros2/rclcpp/blob/b352d450319e84dd28aed78b800ab40938cec5c2/rclcpp_lifecycle/src/node_interfaces/lifecycle_node_interface.cpp#L36-L40) for `on_error` and `on_shutdown` are essentially a noop, I don't see the harm in using `LifecycleNodeInterface` instead (maybe `Node` is a misnomer is this case and the interface could be more appropriately named `LifecycleInterface`).

Reply:
@mkhansen-intel @jacobperron Yes, I agree. If LifecycleNodeInterface became LifecycleInterface, I'd be OK with using LifecycleInterface.  It then wouldn't seem as much of a mis-use of the interface (using a node interface for a "helper").

Reply:
@jacobperron The next issue then becomes how to handle "helper" classes, like TransformBroadcaster, that create publishers. For example, if a TransformBroadcaster is created from a LifecycleNode's topic interface, it must somehow activate any publishers that it creates. Does that mean that all helper classes must support the LifecycleInterface? This imposes a new requirement on non-lifecycle nodes that they must now manage the helper classes. Or, should there be specialized versions of the helper classes, specifically for Lifecycle nodes? Any thoughts here?

Reply:
IMO, objects should be able to create ROS primitives (publishers, subscriptions, services, and actions) using a lifecycle node as they would with a regular node. The only difference being that a lifecycle node has some additional logic for activating/deactivating the ROS primitives. This would remove the burden of managing said helper classes. Although not explicit about implementation, this is the sense that I got from reading the [design doc](http://design.ros2.org/articles/node_lifecycle.html). But I'm not sure if it was the intended plan for lifecycle nodes. @tfoote @gbiggs, maybe you can shed some light on this?

I think this idea could be achieved by implementing a lifecycle variant of the topic and service interfaces and instantiating those [instead of the default](https://github.com/ros2/rclcpp/blob/4f2f8def9819a3e37216254a21e3471655529bc0/rclcpp_lifecycle/src/lifecycle_node.cpp#L65-L66). Then, I *suspect* we might get a desired behavior for actions for free as it is build on top of topics and services, but some more thought it required.